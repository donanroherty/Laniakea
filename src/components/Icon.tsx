/**
 * Icon.tsx
 * Helper class for creating svg icons with props for colour and shadow
 * Icons are defined either by paths in iconPaths or by providing svg child elements
 */

import React from 'react'
import styled from 'theme/themed-styled-components'
import icons, { getIconDimensions } from 'theme/icons'
import shortId from 'shortid'

export interface IconProps {
  fillColor?: string
  size?: number
  icon: keyof typeof icons
}

const defaultProps = {
  fillColor: 'red',
  size: 50,
  icon: 'settings'
}

const Icon = (props: IconProps) => {
  // Get viewbox dimensions from icon definition
  const iconDef = icons[props.icon]
  const vb = iconDef.viewBox
  const vbWidth = vb.x2 - vb.x1
  const vbHeight = vb.y2 - vb.y1

  const dimensions = getIconDimensions(props.icon, props.size!)

  const paths = iconDef.paths.map((val: any, i: any) => (
    <SVGInner
      key={props.icon + 'icon path' + shortId()}
      fillColor={props.fillColor}
    >
      {val}
    </SVGInner>
  ))

  return (
    <SVGWrapper
      {...props}
      viewBox={`0 0 ${vbWidth} ${vbHeight}`}
      height={dimensions.height}
      width={dimensions.width}
    >
      {paths}
    </SVGWrapper>
  )
}

const SVGWrapper = styled.svg<IconProps>`
  pointer-events: none;
  fill: ${props => props.fillColor};
`
const SVGInner = styled.svg<any>`
  fill: ${props => props.fillColor};
`

Icon.defaultProps = defaultProps
export default Icon
