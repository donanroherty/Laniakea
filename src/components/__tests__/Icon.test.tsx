import React from 'react'
import Icon from '../ui/Icon'
import { render, cleanup } from 'react-testing-library'

afterEach(() => cleanup())

test('<Icon />', () => {
  const { container } = render(<Icon icon="settings" />)
  expect(container.firstChild).toMatchSnapshot()
})
