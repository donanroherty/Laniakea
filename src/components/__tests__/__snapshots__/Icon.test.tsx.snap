// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> 1`] = `
<svg
  class="sc-bdVaJa bPmrce"
  height="50"
  size="50"
  viewBox="0 0 100 100"
  width="50"
>
  <svg
    class="sc-bwzfXH QeLOT"
  >
    <path
      class="color1"
      d="M53.3,100h-6.5c-2,0-3.9-1-5-2.8c-0.5-0.7-3.3-5.6-4.8-8.4l-5.2-2.1c-3.2,1-8.9,2.4-9.3,2.5c-2,0.5-4.1-0.1-5.5-1.6
 l-4.6-4.6c-1.4-1.4-2-3.5-1.6-5.5c0.2-0.9,1.6-6.3,2.5-9.3L11.2,63c-3-1.5-8-4.5-8.4-4.8c-1.7-1.1-2.8-3-2.8-5v-6.5c0-2,1-3.9,2.8-5
 c0.7-0.5,5.6-3.3,8.4-4.8l2.1-5.2c-1-3.2-2.4-8.9-2.5-9.3c-0.5-2,0.1-4.1,1.6-5.5l4.6-4.6c1.4-1.4,3.5-2,5.5-1.6
 c0.9,0.2,6.3,1.6,9.3,2.5l5.2-2.1c1.5-3,4.5-8,4.8-8.4c1.1-1.7,3-2.8,5-2.8h6.5c2,0,3.9,1,5,2.8c0.5,0.7,3.3,5.6,4.8,8.4l5.2,2.1
 c3.2-1,8.8-2.4,9.3-2.5c2-0.5,4.1,0.1,5.5,1.6l4.6,4.6c1.4,1.4,2,3.5,1.6,5.5c-0.2,0.9-1.6,6.3-2.5,9.3l2.1,5.2c3,1.5,8,4.5,8.4,4.8
 c1.7,1.1,2.8,3,2.8,5v6.5c0,2-1,3.9-2.8,5c-0.7,0.5-5.6,3.3-8.4,4.8l-2.1,5.2c1,3.2,2.4,8.8,2.5,9.3c0.5,2-0.1,4.1-1.6,5.5l-4.6,4.6
 c-1.4,1.4-3.5,2-5.5,1.6c-0.9-0.2-6.3-1.6-9.3-2.5l-5.2,2.1c-1.5,3-4.5,8-4.8,8.4C57.2,98.9,55.3,100,53.3,100z M31.9,80.7l9.1,3.8
 l0.5,1c0.9,2,4.2,7.6,5,8.8c0.1,0.1,0.2,0.2,0.3,0.2h6.5c0.1,0,0.3-0.1,0.3-0.2c0.6-0.9,4-6.6,5-8.8l0.5-1l9.1-3.8l1,0.4
 c2.1,0.8,8.4,2.4,9.7,2.7c0.1,0,0.3,0,0.3-0.1l4.6-4.6c0.1-0.1,0.1-0.2,0.1-0.3c-0.3-1.1-1.9-7.5-2.7-9.7l-0.4-1l3.7-9.1l1-0.5
 c2-0.9,7.6-4.2,8.8-5c0.1-0.1,0.2-0.2,0.2-0.3v-6.5c0-0.1-0.1-0.3-0.2-0.3l0,0c-0.9-0.6-6.6-4-8.8-5l-1-0.5l-3.8-9.1l0.4-1
 c0.8-2.1,2.4-8.4,2.7-9.7c0-0.1,0-0.3-0.1-0.4l-4.6-4.6c-0.1-0.1-0.2-0.1-0.3-0.1c-1.1,0.3-7.5,1.9-9.7,2.7l-1,0.4L59,15.4l-0.5-1
 c-0.9-2-4.2-7.6-5-8.8c-0.1-0.1-0.2-0.2-0.3-0.2h-6.4c-0.1,0-0.3,0.1-0.3,0.2c-0.6,0.9-4,6.6-5,8.8l-0.5,1l-9.1,3.8l-1-0.4
 c-2.1-0.8-8.4-2.4-9.7-2.7c-0.1,0-0.3,0-0.4,0.1l-4.6,4.6c-0.1,0.1-0.1,0.2-0.1,0.4c0.3,1.1,1.9,7.5,2.7,9.7l0.4,1L15.5,41l-1,0.5
 c-2,0.9-7.6,4.2-8.8,5c-0.1,0.1-0.2,0.2-0.2,0.3v6.5c0,0.1,0.1,0.3,0.2,0.3c0.9,0.6,6.6,4,8.8,5l1,0.5l3.8,9.1l-0.4,1
 c-0.8,2.1-2.4,8.4-2.7,9.7c0,0.1,0,0.3,0.1,0.3l4.6,4.6c0.1,0.1,0.2,0.1,0.4,0.1c1.1-0.3,7.5-1.9,9.7-2.7L31.9,80.7z M50,73.3
 c-12.9,0-23.4-10.5-23.4-23.4S37.1,26.6,50,26.6S73.4,37.1,73.4,50l0,0C73.4,62.9,62.9,73.3,50,73.3z M50,32.2
 c-9.8,0-17.8,8-17.8,17.8s8,17.8,17.8,17.8s17.8-8,17.8-17.8v0C67.8,40.1,59.9,32.2,50,32.2z"
    />
  </svg>
</svg>
`;
